File: problem4.go
package main

import "fmt"
import "strconv"
import "runtime"
import "math"

func IsPalindrome(num int) bool {
	num_str := strconv.FormatInt(int64(num), 10)
	num_str_len := len(num_str)
	for i := 0; i < num_str_len; i++ {
		if num_str[i] != num_str[num_str_len-i-1] {
			return false
		}
	}
	return true
}

func LargestPalindromicProduct(maxDigits int) int {
	highestFound := 0
	for a := 0; a < int(math.Pow10(maxDigits)); a++ {
		for b := 0; b < int(math.Pow10(maxDigits)); b++ {
			product := a * b
			if product > highestFound && IsPalindrome(product) {
				highestFound = product
			}
		}
	}
	return highestFound
}

func SourceLocation(skip int) string {
	_, file, line, ok := runtime.Caller(skip)
	if !ok {
		return "<error>"
	} else {
		return file + ":" + strconv.FormatInt(int64(line), 10)
	}
}

func AssertEq(actual bool, expected bool) {
	if actual != expected {
		fmt.Printf("sanity test failed at %v\n", SourceLocation(2))
		fmt.Printf("  actual: %v\n", actual)
		fmt.Printf("  expected: %v\n", expected)
	}
}

func AssertEqInt(actual int, expected int) {
	if actual != expected {
		fmt.Printf("sanity test failed at %v\n", SourceLocation(2))
		fmt.Printf("  actual: %v\n", actual)
		fmt.Printf("  expected: %v\n", expected)
	}
}

func RunSanityTest() {
	AssertEq(IsPalindrome(1), true)
	AssertEq(IsPalindrome(10), false)
	AssertEq(IsPalindrome(11), true)
	AssertEq(IsPalindrome(12), false)
	AssertEq(IsPalindrome(22), true)
	AssertEq(IsPalindrome(33), true)
	AssertEq(IsPalindrome(110), false)
	AssertEq(IsPalindrome(121), true)
	AssertEq(IsPalindrome(112), false)
	AssertEq(IsPalindrome(121), true)
	AssertEq(IsPalindrome(1111), true)
	AssertEq(IsPalindrome(1221), true)
	AssertEq(IsPalindrome(1231), false)
	AssertEqInt(LargestPalindromicProduct(2), 9009)
	AssertEqInt(LargestPalindromicProduct(3), 906609)
}

func main() {
	RunSanityTest()
	fmt.Printf("answer == %v\n", LargestPalindromicProduct(3))
}

File: problem4.rs
use std::num;

fn is_palindrome(num: int) -> bool {
    let num_str = num.to_str();
    let num_str_len = num_str.len();
    for i in range(0, num_str_len/2) {
        if num_str.as_slice().char_at(i) != num_str.as_slice().char_at(num_str_len-i-1) {
            return false;
        }
    }
    return true;
}

fn largest_palindrome_product(max_digits: uint) -> int {
    let mut highest_found = 0;
    for a in range(1, num::pow(10, max_digits)) {
        for b in range(1, num::pow(10, max_digits)) {
            let prod = a*b;
            if prod > highest_found && is_palindrome(prod) {
                highest_found = prod
            }
        }
    }
    return highest_found
}

#[test]
fn tests() {
    assert_eq!(is_palindrome(0), true);
    assert_eq!(is_palindrome(1), true);
    assert_eq!(is_palindrome(10), false);
    assert_eq!(is_palindrome(11), true);
    assert_eq!(is_palindrome(12), false);
    assert_eq!(is_palindrome(22), true);
    assert_eq!(is_palindrome(33), true);
    assert_eq!(is_palindrome(333), true);
    assert_eq!(is_palindrome(131), true);
    assert_eq!(is_palindrome(1331), true);
    assert_eq!(is_palindrome(1231), false);
    assert_eq!(largest_palindrome_product(2), 9009);
}

fn main() {
    let answer = largest_palindrome_product(2);
    if answer != 9009 {
        fail!("error: wrong answer");
    }
    let answer = largest_palindrome_product(3);
    println!("largest palindrome made from the product of two 3-digit integers is {}", answer);
}
